src/main/java/dk/kb/discover/webservice/KBOAuth2Handler.java

```diff
@@ -0,0 +1,498 @@
+
+        // Note: Timestamps in token is in seconds since Epoch. Date().getTime is milliseconds
+
+        if (trusted.isExpired()) {
+            long overtime = now/1000 - trusted.getExpiration();
+            throw new VerificationException("AccessToken expired (" + overtime + " seconds too old)");
+        }
+
+        if (!trusted.isNotBefore(0)) {
+            long missing = trusted.getNotBefore() - now/1000;
+            throw new VerificationException("AccessToken not valid before " + missing + " seconds has passed");
+        }
+
+        if (trusted.getIssuedAt() > now) {
+            long missing = trusted.getIssuedAt()/1000 - now;
+            log.warn("Received trusted AccessToken issued {}} seconds in the future (epoch seconds = {})",
+                     missing, trusted.getIssuedAt());
+            throw new VerificationException("AccessToken issued " + missing + " seconds in the future");
+        }
+
+        if (!trusted.isActive()) {
+            throw new VerificationException("AccessToken not active, cause not specified");
+        }
+    }
+
+    /**
+     * Extracts the realm from the issuer (iss) and verifies that it is on the list of accepted realms.
+     * @param payload from an access token.
+     * @return the realm, if present and accepted.
+     * @throws VerificationException if the realm cannot be verified.
+     */
+    private String getRealm(JSONObject payload) throws VerificationException {
+        String issuer = payload.getString("iss");
+        Matcher issuerMatcher = ISSUER.matcher(issuer);
+        if (!issuerMatcher.matches()) {
+            String error = "Unable to determine realm from token payload iss '" + issuer + "'";
+            log.warn(error);
+            throw new VerificationException(error);
+        }
+
+        String tokenRealm = issuerMatcher.group(2);
+        if (!realms.contains(tokenRealm)) {
+            String error = String.format(
+                    Locale.ROOT, "The provided realm '%s' from iss '%s' was not on the list of allowed realms",
+                    tokenRealm, issuer);
+            log.warn(error + " " + realms);
+            throw new VerificationException(error);
+        }
+        return tokenRealm;
+    }
+    private static final Pattern ISSUER = Pattern.compile("^(.*)/(.+)/?$");
+
+    private JSONObject decodeJSONObject(String base64JSON) throws VerificationException {
+        String jsonString = new String(base64Decode(base64JSON), StandardCharsets.UTF_8);
+        try {
+            return new JSONObject(new JSONTokener(jsonString));
+        } catch (Exception e) {
+            log.warn("Unable to parse JSON in access token '{}'", jsonString, e);
+            throw new VerificationException("Unable to parse JSON in access token");
+        }
+    }
+
+    // The Base64 strings that come from a JWKS need some manipulation before they can be decoded.
+     // we do that here
+     private static byte[] base64Decode(String base64) {
```

Comment:
Please convert to correct javadoc and maybe explain in a bit more detail what is getting manipulated

Reply:
Very good point. Seems very suspicious that should be required.
Fixed.. With a 'wtf' comment added... (another TODO)